mountpoint-s3/src/prefetch/caching_stream.rs

```diff
@@ -229,11 +283,39 @@ where
 
             // Split the body into blocks.
             let mut body: Bytes = body.into();
+            let mut offset = offset;
             while !body.is_empty() {
                 let remaining = (block_size as usize).saturating_sub(self.buffer.len()).min(body.len());
-                let chunk = body.split_to(remaining);
+                let chunk: ChecksummedBytes = body.split_to(remaining).into();
+
+                // We need to return some bytes to the part queue even before we can fill an entire caching block because
+                // we want to start the feedback loop for the flow-control window.
+                //
+                // This is because the read window might be enough to fetch "some data" from S3 but not the entire block.
```

Comment:
Not sure I follow. Are we aligning to block boundaries or not?
Performance-wise, it would be a different trade-off, happy to discuss it. But we need to make sure we do not try to cache partial blocks.

Reply:
Sorry, I mean the request ranges are aligned to block boundaries, but the read window ranges may not. Agree, that we must not try to cache partial blocks.